Adding latent variables covariance model for categorical variables

JPelamatti opened the pull request:
This is a PR for the implementation of a covariance model allowing to compute the covariance between the levels of a discrete/categorical variables based on the mapping onto a d-dimensional latent space.

If you are interested in a reference explaining the theory : https://arxiv.org/abs/1806.07504

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,360 @@
+//                                               -*- C++ -*-
+/**
+ *
+ *  Copyright 2005-2022 Airbus-EDF-IMACS-ONERA-Phimeca
```
same for header

sofianehaddad reviewed (commented):
I am just starting to review, not finished
But have a couple of questions; probably needs a deep eye on the article firstly

JPelamatti commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,360 @@
+  activeLatentCoordinateDim_ = 1 + latent_dim_ * (n_levels_ - 2);
+  activeLatentVariables_ = Point(activeLatentCoordinateDim_,0.);
+  fullLatentVariables_ = Sample(n_levels_,latent_dim_);
+  Indices activeParameter = Indices(inputDimension_ + outputDimension_ + activeLatentCoordinateDim_);
```
Thanks for the review!
I'm not sure I understand the first comment.
Concerning the second comment, yes input and output dimension are 1 by default and cannot be modified. Should I replace this line by : 
`Indices activeParameter = Indices(2 + activeLatentCoordinateDim_);` ?

sofianehaddad replied:
Oh I was confusing `latent dim` & `active parameter`'s size

JPelamatti commented:
The compilation of the documentation seems to call the 'draw' method of the LatentVariableModel class, which I have not overridden, and therefore generates a sphinx warning. 
In this case, the draw method should return a short of checkerboard plot, where each cell i,j is characterized by the covariance value between level i and level j. Is this sort of plot doable in OpenTURNS? And in the meanwhile, is it possible to make it so that the sphinx compilation does not try to draw this covariance function?

jschueller commented:
check the pyplot scripts

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (*z1_it == i) isLevelz1 = true;
```
missing break

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+  // Set the full sample of latent coordinates
+  // Fix the coordinates of the first two latent variables
+  fullLatentVariables_[0] = Point(latent_dim_,0.);
+  Point levelMapping = Point(latent_dim_,0.);
```
```suggestion
 Point levelMapping(latent_dim_,0.);
```

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+  fullLatentVariables_[0] = Point(latent_dim_,0.);
+  Point levelMapping = Point(latent_dim_,0.);
+  levelMapping[0] = latentVariablesCoordinates[0];
+  fullLatentVariables_[1] = levelMapping;
```
```suggestion
  for (UnsignedInteger k = 0; k < latent_dim_; ++k) fullLatentVariables_(1, k) = levelMapping[k]; // avoid creating latent_dim_ points
```

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+      levelMapping[j] = latentVariablesCoordinates[count];
+      count++;
+    }
+    fullLatentVariables_[i] = levelMapping;
```
```suggestion
    for (UnsignedInteger k = 0; k < latent_dim_; ++k) fullLatentVariables_(i, k) = levelMapping[k]; 
```

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (z1[0] == i) isLevelz1 = true;
```
missing break

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (z2[0] == i) isLevelz2 = true;
```
missing break

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+
+Scalar LatentVariableModel::computeAsScalar(const Point & z1, const Point & z2) const
+{
+  /*check z being one of the levels*/
```
As you are 1D you could call `computeAsScalar(z1[0], z2[0])`

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,455 @@
+
+  for (UnsignedInteger i = 0; i < n_levels_; ++i)
+  {
+    if (*z2_it == i) isLevelz2 = true;
```
missing break

sofianehaddad reviewed (commented):
Still reviewing

JPelamatti commented:
I have overridden the draw method (although it's a temporary fix), but the doc compiler still seems to call the parent version of said method (which requires the computation of computeAsScalar(t,s), not doable on a given range 0,1 in our case).  I am fairly confused.

jschueller commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,137 @@
+  UnsignedInteger latent_dim_;
+
+  // The number of levels
+  UnsignedInteger n_levels_;
```
use camelcase

jschueller commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+  
+  // Dummy line in order to use all inputs
+  Scalar dummyVariable = 0.; 
+  if (asStationary && correlationFlag) dummyVariable = zMin+zMax;  
```
just comment the unused variables: `..., const Bool /*correlationFlag*/) const`

jschueller commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,137 @@
+
+  /** Parameters constructor */
+  LatentVariableModel(const UnsignedInteger latent_dim,
+						 const UnsignedInteger levels);
```
the order of variables is confusing here, put the most important one first then keep the same order.

sofianehaddad replied:
Indeed

jschueller commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+
+  Scalar cov = latCovMat_(*z1_it,*z2_it);
+  return cov;
+}
```
if the model is 1-d only you can maybe avoid to duplicate code and call the previous version

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+  {
+    for (UnsignedInteger j = 0; j <= i; ++j)
+    {
+      latCovMat_(i,j) = latCovMod_(fullLatentVariables_[i], fullLatentVariables_[j])(0,0);
```
Some remarks/questions:
 - use computeAsScalar instead `latCovMod_.computeAsScalar(fullLatentVariables_[i], fullLatentVariables_[j]);`
 - Do we really need to embed a `SquaredExponential` to compute `exp(-0.5(fullLatentVariables_[i], fullLatentVariables_[j])^2` ?
 - In the article there is no 1/2 factor in the exp function. So ...

sofianehaddad commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,137 @@
+  UnsignedInteger activeLatentCoordinateDim_;
+
+  // The active latent variable coordinates parameter
+  Point activeLatentVariables_;
```
name is confusing. `latenttVariablesCoordinates_` ?

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+
+  // Set the full sample of latent coordinates
+  // Fix the coordinates of the first two latent variables
+  fullLatentVariables_ = Sample(n_levels_, Point(latent_dim_,0.));
```
```suggestion
  fullLatentVariables_ = Sample(n_levels_, latent_dim_);
```

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+    throw InvalidArgumentException(HERE) << "In LatentVariableModel::setLatentVariables, samples have incompatible size. Sample size = " << latentVariablesCoordinates.getSize()
+                                         << " whereas expected size = " << activeLatentCoordinateDim_;
+  // Set the active variables
+  activeLatentVariables_ = latentVariablesCoordinates;
```
the name `activeLatentVariables_` is confusing

mbaudin47 replied:
@sofianehaddad What name would you suggest?

JPelamatti commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+  {
+    for (UnsignedInteger j = 0; j <= i; ++j)
+    {
+      latCovMat_(i,j) = latCovMod_(fullLatentVariables_[i], fullLatentVariables_[j])(0,0);
```
I thought about it, but I figured this could be faster given that the latCovMat_ matrix should rarely be large enough for the 'discretize' parallelization to be very efficient, and this way we skip all the unnecessary checks on the method inputs. 

It's true that the discretize call makes for a more readable code.

sofianehaddad replied:
yeah, more readable on the one hand. On the other hand, I don't know about `n_levels_` values but I think parallelization will not be efficient and will not downgrade performance. If so we might implement a `sequential discretize`. We keep in mind that in an optim loop, there is no concurrency with `LatentVariableModel.discretize`

sofianehaddad commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,441 @@
+      fullLatentVariables_(i,j) = latentVariablesCoordinates[count];
+      count++;
+    }
+  }
```
it is difficult/hard to read, understand, maintain.

I understood that `activeLatentVariables_` is of size `n_dim_ * (n_levels_ - 2) + 1` or `n_levels_ * n_dim_ - 2 * n_dim_ + 1`, and sample size is `n_levels_ * n_dim_`, butt why is the sample filled like this?

jschueller commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,50 @@
+Available constructors:
+    LatentVariableModel(*nlevels=2*)
+
+    LatentVariableModel(*latentDim, nlevels*)
```
order is wrong

jschueller commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,50 @@
+nlevels : int
+    Number of levels :math:`l` characterizing the discrete variable.
+latentDim : int
+	The dimension :math:`d_l` of the latent space onto which the discrete variable levels are projected.
```
indent

jschueller commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,50 @@
+
+Notes
+-----
+The *Latant variable covariance function* is a covariance model allowing to compute the covariance between different unordered values (or levels) of a discrete variable.
```
Latant

jschueller commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,50 @@
+
+The coordinates of the mapping points are part of the covariance model parameters, together with the latent squared exponential model scale and amplitude.
+it is important to note that in order to compensate for possible rotations and translations of the mapping points, the coordinates of the first level mapping
+are fixed to the latent space origin, whereas all of the coordinates of the second level mapping are fixed to :math:`0`, except for the first one. 
```
shorter lines

JPelamatti commented:
It's strange, I obtain considerably better results for the example running it on my local install than what is obtained in the compiled doc.

sofianehaddad commented:
> It's strange, I obtain considerably better results for the example running it on my local install than what is obtained in the compiled doc.

I had currently tthe same with my PR!

JPelamatti commented:
I added a second covariance model, simpler and more easily understandable, although less flexible. You can see the performance of the two alternatives in the new example

JPelamatti commented:
Waiting for a live discussion, I have removed the second, simpler, covariance kernel from this PR.
I have also removed the draw method which is not necessary and which still requires some additional work.

jschueller commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,137 @@
+
+  /** Parameters constructor */
+  LatentVariableModel(const UnsignedInteger nLevels,
+						 const UnsignedInteger latentDim);
```
both nLevels & latentDim could have default values so only one ctor is needed:
LatentVariableModel(const UnsignedInteger nLevels = 2, const UnsignedInteger latentDim = 2);

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+ot.RandomGenerator.SetSeed(5)
```
When the library is loaded, the seed is automatically set to zero. Setting the seed to 5 is either a choice that is unnecessary (because removing it generates an interesting example) or deserves a comment to explain its purpose.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+# LatentVariableModel for the discrete ones.
+
+# %%
+latdim = 1  # Dimension of the latent space
```
The variable name `latdim` hides its meaning. I suggest to name it `latentDimension`, so that the comment is unnecessary:
```suggestion
latentDimension = 1
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+
+
+# %%
+def illustrativeFunc(inp):
```
This function name  could be chosen in every example in the doc. What is the specific purpose of this particular function?

mbaudin47 replied:
Is that function used in the paper (Zhang et al., 2020)?

JPelamatti replied:
The two test functions are from the PhD thesis of J. Pelamatti. A reference has been added.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+# %%
+def illustrativeFunc(inp):
+    x, z = inp
+    x = 7 * x
+    y = np.cos(x) + 0.5 * z
```
```suggestion
    y = np.cos(7 * x) + 0.5 * z
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+lvls = 2  # Number of discrete levels for z
+# Input distribution
+dist = ot.ComposedDistribution(
+    [ot.Uniform(0, 1), ot.UserDefined(np.arange(0, lvls).reshape((-1, 1)))]
```
This is an example in OT, not NP:
```suggestion
    [ot.Uniform(0, 1), ot.UserDefined(ot.Sample.BuildFromPoint(range(lvls)))]
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+
+dim = 2
+fun = ot.PythonFunction(dim, 1, illustrativeFunc)
+lvls = 2  # Number of discrete levels for z
```
Using a variable which explains its content produces a simpler code:
```suggestion
numberOfZLevels = 2
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+)
+
+# %%
+# In this example, we compare the performances of the LatentVariableModel with a naive
```
```suggestion
# In this example, we compare the performances of the :class:`~openturns.LatentVariableModel` with a naive
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+
+# %%
+# In order to deal with mixed continuous / discrete problems we can rely on the
+# ProductCovarianceModel class. We start here by defining the product kernel,
```
```suggestion
# :class:`~openturns.ProductCovarianceModel` class. We start here by defining the product kernel,
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,322 @@
+  if (nLevels_ < 2) throw InvalidArgumentException(HERE) << "Error: the number of discrete levels must be >= 2";
+  activeLatentCoordinateDim_ = 1 + latentDim_ * (nLevels_ - 2);
+  activeLatentVariables_ = Point(activeLatentCoordinateDim_,0.);
+  fullLatentVariables_ = Sample(nLevels_,latentDim_);
```
```suggestion
  fullLatentVariables_ = Sample(nLevels_, latentDim_);
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+# One for each combination of categorical variables levels.
+
+# %%
+k_ind = ot.SquaredExponential(1)
```
OT uses CamelCase (and not the snake_case common in Python), so I suggest to use this consistently in the examples.
```suggestion
kIndex = ot.SquaredExponential(1)
```
The same comment applies to many other variables below, including `lb_ind`, `bounds_ind`, etc.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+
+# %%
+# We generate the training data set
+X = dist.getSample(10)
```
An upper case first letter should be used for class names only.
![image](https://github.com/openturns/openturns/assets/31351465/3f35e2de-ea05-4415-b27f-3d2d0f90fb8d)

```suggestion
xSample = dist.getSample(10)
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+for z in range(2):
+    # We select the training samples corresponding to the correct combination
+    # of categorical levels
+    ind = np.where(np.all(np.array(X[:, 1]) == z, axis=1))[0]
```
The `Sample.find` [method](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.Sample.html#openturns.Sample.find) does the job of finding a particular value in the `X` sample.

JPelamatti replied:
Unfortunately the Sample.find method only returns the index of the first value encountered, here we need to recover all the indices where the second column of x is equal to z.

mbaudin47 replied:
OK, you are right. I thought that the `[0]` only get the first match.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Is there any bibliographical reference for this function?

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Idem.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+# %%
+# We plot the prediction of the mixed continuous / categorical GP,
+# as well as the one of the two separate continuous GPs
+fig, (ax1, ax2) = plt.subplots(2, 1, sharex=True, figsize=(15, 10))
```
This should be done using [GridLayout](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.GridLayout.html).

mbaudin47 replied:
As far as I understand, you do not want to use the graphics OpenTURNS API within Python examples. What not to discuss that topic at the next technical meeting?

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+for z in range(2):
+    # We select the training samples corresponding to the correct combination
+    # of categorical levels
+    ind = np.where(np.all(np.array(X[:, 1]) == z, axis=1))[0]
```
Idem: please use `Sample.find()`.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+
+    # Compute the models predictive performances on a validation data set
+    # We compute the predictions independently for each level of z
+    ind = np.where(np.all(np.array(Xplt[:, 1]) == z, axis=1))[0]
```
Idem: please use `Sample.find()`.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+        algo_ind_list[z].getConditionalMarginalVariance(Xplt_ind[:, 0])
+    )
+
+    (trainingData,) = ax1.plot(Xl[:, 0], Yl, "r*")
```
Please use [Cloud](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.Cloud.html#openturns.Cloud).

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```diff
@@ -0,0 +1,375 @@
+    ax2.plot(Xl[:, 0], Yl, "r*")
+    ax2.plot(Xplt_ind[:, 0], Yplt_ind, "k--")
+    ax2.plot(Xplt_ind[:, 0], predMeanInd, "b-")
+    ax2.fill_between(
```
This is shown [here](http://openturns.github.io/openturns/latest/auto_graphs/plot_graphs_fill_area.html) in OT.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Idem: please choose a proper variable name, including the camel case.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
```suggestion
ub_LV = [3.0] * dim + [10.0] * activecoord
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Please use [Black](https://pypi.org/project/black/) on your script:
```suggestion
ub_ind = [3.0, 3.0]
```

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Up to the change from $n$ to $n - 1$, this looks like [Sample.computeStandardDeviation()](http://openturns.github.io/openturns/latest/user_manual/_generated/openturns.Sample.html#openturns.Sample.computeStandardDeviation), isn'it?

JPelamatti replied:
It is similar, but on a finite sample the average residual will not be null.

mbaudin47 replied:
I do not understand your answer. Moreover, I guess that this reveals that a new MetaModelValidation.computeRootMeanSquaredError() method should be created. This is out of the scope of this PR though.

mbaudin47 replied:
I created MetaModelValidation.computeMeanSquaredError() at https://github.com/openturns/openturns/pull/2330/files#diff-8eb6060d1ee247fcf98f5f7fcb0a59655266d95063bb3f27f0d58f1d4d30531d. Should I create MetaModelValidation.computeRootMeanSquaredError() so that this particular example can be simplified?

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Idem: please use `Sample.find()`.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Idem: please use `Sample.find()`.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Idem.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
This looks inappropriate for two reasons:
- the number of entries in the array is known in advance,
- `Sample` should be used instead.

mbaudin47 commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
Is the magic number 3 equal to `1 + latdim`?

JPelamatti replied:
actually the two 3s are numberOfZLevels1 and numberOfZLevels2, they have been replaced

mbaudin47 reviewed (changes requested):
This is an interesting new covariance model: thank you for this work. It looks to me that a small work in the example would make it go from a Numpy / Matplotlib example into an OpenTURNS example.

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+As a result, the number of active latent variable coordinates is equal to :math:`d_l * (l - 2) + 1`. In practice, 
+the class distinguishes between the **FullLatentVariables** attribute, which contains the actual latent variables
+coordinates, and the **ActiveLatentVariables** attribute, which contains only the coordinate which can be modified.
```
This makes the comment consistent with the API help page presented below:
```suggestion
coordinates, and the `activeLatentVariables` attribute, which contains only the coordinate which can be modified.
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+origin, whereas all of the coordinates of the second level mapping are fixed to :math:`0`, except for the first one. 
+
+As a result, the number of active latent variable coordinates is equal to :math:`d_l * (l - 2) + 1`. In practice, 
+the class distinguishes between the **FullLatentVariables** attribute, which contains the actual latent variables
```
This makes the comment consistent with the API help page presented below:
```suggestion
the class distinguishes between the `fullLatentVariables` attribute, which contains the actual latent variables
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+Is is important to note that for the sake of simplicity, the discrete variable levels **must be** represented as integers, 
+ranging from :math:`0` to :math:`l-1`. However, this representation is purely practical, and the values
+assigned to each level have no practical meaning or effect.
```
```suggestion
assigned to each level have no practical meaning or effect: only the coordinates matter in the
value of the covariance.
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+assigned to each level have no practical meaning or effect.
+
+Finally, for a similar reason, when using this type of kernel when defining a Gaussian process, it is suggested to 
+rely on a constant functional basis.
```
```suggestion
rely on a constant functional basis: please use the :class:`~openturns.ConstantBasisFactory` class.
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+%feature("docstring") OT::LatentVariableModel::computeAsScalar
+"Compute the covariance function for scalar model.
+
+Available usages:
+    computeAsScalar(s, t)
```
This will be done by Sphinx, hence the code should not be there:
```suggestion
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+Parameters
+----------
+s, t : floats, must have integer values between :math:`0` and :math:`l-1`
```
This produces:
![image](https://github.com/openturns/openturns/assets/31351465/595eeef0-50be-4eb4-8373-56a576056813)
Please look at [Sample](https://github.com/openturns/openturns/blob/650bb5e463f62d5e5519615655e33db87d2ca52c/python/src/Sample_doc.i.in#L282) for examples.


```suggestion
s, t : floats
    Must have integer values between :math:`0` and :math:`l-1`
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+Notes
+-----
+The method makes sense only if the dimension of the process is :math:`d=1`.
+It evaluates :math:`C(s, t)`."
```
... but what exactly is $C$ in this method?

JPelamatti replied:
It is the covariance of the process, with the difference that the process is not defined over time, but over a categorical discrete space

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+activeLatentVariables : :class:`~openturns.Point`
+    Active coordinates of the discrete levels in the latent space. 
+    The inactive coordinates are set to 0 (i.e., the first latent 
+    variable is placed in the Euclidean space origin, and the second
```
Is "placed" the correct word here? @josephmure surely knows would to turn this into proper english.

mbaudin47 replied:
I think "located" is better.

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+    Active coordinates of the discrete levels in the latent space. 
+    The inactive coordinates are set to 0 (i.e., the first latent 
+    variable is placed in the Euclidean space origin, and the second
+    latent variable can only be placec onto the first axis)."
```
```suggestion
    latent variable can only be placed onto the first axis)."
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+Returns
+-------
+LatentDimension : int
```
Only classes names have an upper case first letter.
```suggestion
latentDimension : int
```

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+Returns
+-------
+NLevels : int
```
idem.

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+
+Parameters
+----------
+LatentVariables : :class:`~openturns.Point`
```
- idem on the upper case first letter
- SWIG automatically converts any sequence of floats into a `Point`.

This is why the sented `sequence of float` is consistently used in the doc (e.g. in the doc of [Sample.add()](https://github.com/openturns/openturns/blob/650bb5e463f62d5e5519615655e33db87d2ca52c/python/src/Sample_doc.i.in#L210)).

mbaudin47 commented on python/src/LatentVariableModel_doc.i.in:
```diff
@@ -0,0 +1,144 @@
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::LatentVariableModel::draw
+"Not yet implemented"
```
Is that necessary?

JPelamatti replied:
I believe it is necessary, as the parent class provides a documented implementation of the draw method.

mbaudin47 commented on python/src/experimental_module.i.orig:
```diff
@@ -0,0 +1,71 @@
+// SWIG file experimental_module.i
```
This looks that a git merge file problem here: this file should probably not be here.

mbaudin47 commented on python/test/t_LatentVariableModel_std.py:
```diff
@@ -0,0 +1,18 @@
+k.setLatentVariables([0.1, 0.2, 0.3, -0.1, -0.2, -0.3, 0.4])
+k.setScale([1.5])
+k.setAmplitude([2.])
+ott.assert_almost_equal(k(1, 1)[0, 0], 4.000000000004)
```
idem as in C++

mbaudin47 commented on python/test/t_LatentVariableModel_std.py:
```diff
@@ -0,0 +1,18 @@
+import openturns.testing as ott
+
+ot.TESTPREAMBLE()
+ot.RandomGenerator.SetSeed(0)
```
This is unnecessary.

jschueller replied:
just drop setseed

mbaudin47 commented:
In order to make the PR easier to review, I suggest to briefly introduce the new class: its name, its main properties and the issues that may be raised from its implementation (but do not see any).

m-balesdent reviewed (changes requested):
Thank you very much @JPelamatti for this PR that provide very interesting features on the surrogate modeling.

JPelamatti commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,322 @@
+    }
+  }
+
+  if (!isLevelz1 || !isLevelz2) throw InvalidArgumentException(HERE) << "Error: the input discrete variables values are not amongst the known levels";
```
In my opinion, a zero covariance has a very specific meaning, whereas the error here is raised when the user is providing categorical levels which do not exist, i.e., the user is asking for a covariance value that carries no meaning

JPelamatti commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,134 @@
+  CovarianceMatrix latCovMat_;
+
+  // The latent covariance model
+  SquaredExponential latCovMod_;
```
Yes in theory it is absolutely doable, but I opted for a simpler API for this first implementation

mbaudin47 replied:
Does this provide a covariance model whatever the choice of the root covariance model? I.e. is the covariance matrix always definite positive?

JPelamatti replied:
If the root covariance model is SDP, so is the resulting categorical covariance model, as it ends up simply being an evaluation of the root covariance model on a finite set of points in R^d. It must however be noted that the resulting covariance matrix will very often be non-inversible, as the input sample will contain repeated values.

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,322 @@
+Scalar LatentVariableModel::computeAsScalar(const Scalar z1, const Scalar z2) const
+{
+  /*check z being one of the levels*/
+  bool isLevelz1 = false;
```
I did not see that point. In that case, I suggest to check that the fractional part of z1 is zero. You can do that by evaluating the difference between that number and its cast into an integer. If the difference is non zero, then an exception can be raised. This is an explicit implementation of your requirement, but it avoids a `for` loop.

JPelamatti commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,134 @@
+  Point getFullParameter() const override;
+
+  /** Scale accessors */
+  void setScale(const Point & scale) override;
```
Unfortunately, these two method names must be consistent with the parent class name which is overridden.

JPelamatti commented on lib/src/Base/Stat/openturns/LatentVariableModel.hxx:
```diff
@@ -0,0 +1,134 @@
+
+  /** Parameters constructor */
+  LatentVariableModel(const UnsignedInteger nLevels = 2,
+						 const UnsignedInteger latentDim = 2);
```
The input and output dimension can only be 1 for this covariance model, which is consistent with the default constructors of the common covariance models in OT

mbaudin47 replied:
Ok. But the code still has to be indented with blank spaces. Perhaps you may use another text editor to see the difference?

jschueller commented on ChangeLog:
```diff
@@ -37,6 +37,7 @@
  * LeastSquaresExpansion, IntegrationExpansion (openturns.experimental)
  * UniformOverMesh (openturns.experimental)
  * GeneralizedExtremeValueValidation (openturns.experimental)
+ * LatentVariableModel (openturns.experimental)
```
should go into the 1.22 section (rebase if needed)

jschueller commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,300 @@
+  adv.saveAttribute("activeLatentVariables_", activeLatentVariables_);
+  adv.saveAttribute("nLevels_", nLevels_);
+  adv.saveAttribute("latentDim_", latentDim_);
+// ADD OTHER ATTRIBUTES
```
which ones ?

jschueller commented on python/src/experimental_module.i.orig:
```diff
@@ -0,0 +1,71 @@
+%include LatentVariableModel
+=======
+%include LatentVariableModel.i
+>>>>>>> c961cdf312afc4c92a768c87c37a102e3a49ed86
```
probably a unresolved conflict

JPelamatti commented on python/doc/examples/meta_modeling/kriging_metamodel/plot_kriging_categorical.py:
I personally prefer using python functions in the examples, except when the performance of a symbolic function is required, as I believe it is a more common usage of OT, and thus easier to understand / relate to.

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+static const Factory<LatentVariableModel> Factory_LatentVariableModel;
+
+/** Parameters constructor */
+LatentVariableModel::LatentVariableModel(const UnsignedInteger nLevels,
+										 const UnsignedInteger latentDim)
```
The tabulations are an unstable way to indent the code: please use blank spaces.
```suggestion
LatentVariableModel::LatentVariableModel(const UnsignedInteger nLevels,
                                         const UnsignedInteger latentDim)
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  const bool isLevelz2 = (z2 >= 0 && z2 < nLevels_ && (z2 - floor(z2)) == 0);
+
+  if (!isLevelz1 || !isLevelz2) throw InvalidArgumentException(HERE) << "Error: the input discrete variables values: " << z1 
+  << ", and/or " << z2 << " are not among the known levels. They should both present integer values between 0 and l-1.";
```
I guess that the "<<" continues the printing, but the indentation makes it confusing.
```suggestion
    << ", and/or " << z2 << " are not among the known levels. They should both present integer values between 0 and l-1.";
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,322 @@
+
+void LatentVariableModel::setFullParameter(const Point & parameter)
+{
+  /*
```
I notice that the line 104 check that the size of the `parameter` input argument is correct. Should that comment be moved into the API doc? Perhaps I missed that and the API doc already contains the information.

JPelamatti replied:
Fixed, the API for set/getFullParameter and set/getParameter has been overridden with this additional information

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  // Check the size
+  const UnsignedInteger totalSize = 2 + activeLatentCoordinateDim_;
+  if (!(parameter.getSize() == totalSize))
+    throw InvalidArgumentException(HERE) << "In LatentVariableModel::setFullParameter, points have incompatible size. Point size = " << parameter.getSize()
```
The information within the message is very good.
I think, however, that the macro `HERE` already gives the location of the error within the call stack, isn'it? So I guess that the name of the method `LatentVariableModel::setFullParameter` is not necessary in the message itself.
```suggestion
    throw InvalidArgumentException(HERE) << "The parameter has size = " << parameter.getSize()
```
Perhaps you could write a Python script which generates the exception and show the printed message as a reply to this discussion so that we can see what is actually printed.

JPelamatti replied:
With the current version of the code, we have:

```
import openturns.experimental as otexp
kLat = otexp.LatentVariableModel(2,2)
kLat.setFullParameter([1.]*10)
```

output:
_In LatentVariableModel::setFullParameter, points have incompatible size. Point size = 10 whereas expected size = 3_

Therefore I am guessing the macro HERE does not behave like you think

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  return computeAsScalar(*z1_it, *z2_it);
+}
+
+
```
At this point, a one-line comment of the method is mandatory, as for all C++ methods in the cxx.
```suggestion
/* Set the vector of parameters */
```
Please add this for all methods below: some have this comment, but many do not.

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  setLatentVariables(activeLatentVariables);
+}
+
+Point LatentVariableModel::getFullParameter() const
```
```suggestion
/* Return the vector of parameters */
Point LatentVariableModel::getFullParameter() const
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  return parameter;
+}
+
+Description LatentVariableModel::getFullParameterDescription() const
```
```suggestion
/* Accessor to the description of the vector of parameters */
Description LatentVariableModel::getFullParameterDescription() const
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  updateLatentCovarianceMatrix();
+}
+
+void LatentVariableModel::updateLatentCovarianceMatrix()
```
```suggestion
/* Update the covariance matrix of latent variables */
void LatentVariableModel::updateLatentCovarianceMatrix()
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  return fullLatentVariables_;
+}
+
+Point LatentVariableModel::getActiveLatentVariables() const
```
```suggestion
/* Accessor to the active latent variables */
Point LatentVariableModel::getActiveLatentVariables() const
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  return nLevels_;
+}
+
+void LatentVariableModel::setScale(const Point & scale)
```
```suggestion
/* Set the vector of scales */
void LatentVariableModel::setScale(const Point & scale)
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  updateLatentCovarianceMatrix();
+}
+
+void LatentVariableModel::setAmplitude(const Point & amplitude)
```
```suggestion
/* Set the vector of amplitude */
void LatentVariableModel::setAmplitude(const Point & amplitude)
```

mbaudin47 commented on lib/src/Base/Stat/LatentVariableModel.cxx:
```diff
@@ -0,0 +1,299 @@
+  updateLatentCovarianceMatrix();
+}
+
+void LatentVariableModel::setNuggetFactor(const Scalar nuggetFactor)
```
```suggestion
/* Set the nugger factor */
void LatentVariableModel::setNuggetFactor(const Scalar nuggetFactor)
```

mbaudin47 commented on lib/test/t_LatentVariableModel_std.cxx:
```diff
@@ -0,0 +1,49 @@
+	k.setScale(Point({1.5}));
+	k.setAmplitude(Point({2.0}));
+	assert_almost_equal(k(1, 1)(0, 0), 4.0);
+	assert_almost_equal(k(1, 2)(0, 0), 3.903408);
```
Where does the magic constant 3.903408 come from? If this is a computation, you might compute the reference value from the C++. This question is because maintaining that code may be difficult for another developer who does not have this information.

mbaudin47 reviewed (changes requested):
Many comments already have been fixed, but there is still room for improvement.

JPelamatti commented:
All comments have been taken into account (I think), however the doc does not compile for a mystical reason that eludes me.

jschueller commented:
you need to rebase

JPelamatti commented:
magic

mbaudin47 commented:
@m-balesdent Are you OK with the changes done by @JPelamatti or is there any room for improvement from your point of view?

jschueller commented:
a rebase is needed (& adress the last few comments)

jschueller commented:
a rebase is needed

mbaudin47 commented:
@JPelamatti : Please rebase so that the code can be merged.

JPelamatti commented:
Most comments and suggestions have been taken into account. 
The only issues remaining are the naming of some of the methods and attribute, however I could not come up with something more explicit.
It is my opinion that this PR can be merged as it only contains an 'experimental' class, and having it in the 1.22 version could help finetune the remaining naming and user friendliness issues.

jschueller commented:
yep, maybe @sofianehaddad or @regislebrun could approve and I'll merge it

mbaudin47 commented on ChangeLog:
```diff
@@ -93,9 +94,12 @@
  * Add VisualTest::DrawPPplot
  * Add VisualTest.Draw(Upper|Lower)(Tail|Extremal)DependenceFunction methods to plot dependence functions
  * Add Distribution.compute(Upper|Lower)(Tail|Extremal)DependenceMatrix methods to compute dependence coefficients
+<<<<<<< HEAD
```
```suggestion
```

mbaudin47 commented on ChangeLog:
```diff
@@ -93,9 +94,12 @@
  * Enable Pagmo.moead_gen with pagmo>=2.19
  * Enable Bonmin.Ecp/iFP algorithms with bonmin>=1.8.9
  * BoxCoxFactory handles linear model
+=======
+>>>>>>> 6873d3bc6 (corrected changelog)
```
```suggestion
```

mbaudin47 commented on ChangeLog:
```diff
@@ -122,6 +126,7 @@
  * #2281 (The marginal of FireSatelliteModel can fail)
  * #2285 (The input and output descriptions does not go down to the metamodel)
  * #2287 (DesignProxy.computeDesign() can produce a segmentation fault)
+<<<<<<< HEAD
```
```suggestion
```

mbaudin47 commented on ChangeLog:
```diff
@@ -130,6 +135,11 @@
  * #2327 (A sign mistake in FGM document)
  * #2328 (Sample.add weird behaviour)
  * #2338 (HSIC with large amount of data makes crash)
+=======
```
```suggestion
```
